File: algorithms/0014/main.go
package p0014

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	shortestStr := strs[0]
	for _, str := range strs {
		if len(str) < len(shortestStr) {
			shortestStr = str
		}
	}

	for i := 0; i < len(shortestStr); i++ {
		for j := 0; j < len(strs); j++ {
			if strs[j][i] != shortestStr[i] {
				return shortestStr[:i]
			}
		}
	}
	return shortestStr
}

File: algorithms/0014/main.js
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function(strs) {
  // 排除个别情况
  if (strs.length < 2) {
    return strs[0] || ''
  }
  // 找到最短的字符串
  const shortestStr = strs.reduce((acc, cur, i) => {
    if (cur.length < acc.length) {
      acc = cur
    }
    return acc
  }, strs[0])
  for (let i = 0; i < shortestStr.length; i++) {
    const isEqual = strs.every(str => str[i] === shortestStr[i])
    if (!isEqual) {
      return shortestStr.slice(0, i)
    }
  }
  return shortestStr
}

module.exports = longestCommonPrefix
